diff --git a/node_modules/@embroider/shared-internals/src/ember-standard-modules.js b/node_modules/@embroider/shared-internals/src/ember-standard-modules.js
index 0b74c5a..0abd658 100644
--- a/node_modules/@embroider/shared-internals/src/ember-standard-modules.js
+++ b/node_modules/@embroider/shared-internals/src/ember-standard-modules.js
@@ -34,6 +34,11 @@ exports.emberVirtualPeerDeps.add('@ember/string');
 exports.emberVirtualPackages.add('@glimmer/env');
 exports.emberVirtualPackages.add('ember');
 exports.emberVirtualPackages.add('@embroider/macros');
+
+exports.emberVirtualPackages.add('@glimmer/manager');
+exports.emberVirtualPackages.add('@ember/owner');
+exports.emberVirtualPackages.add('@ember/renderer');
+
 // These are the known names that people use to import template precomiplation
 // macros from.
 exports.templateCompilationModules = Object.freeze([
